sql: return a copy of array in arrayAggregate

yuzefovich opened the pull request:
Makes arrayAggregate follow its contract by returning
a copy of accumulated array with every call to Result().

Fixes: #28287.

Release note (bug fix): CockroachDB now correctly
handles computation of `array_agg` when used as
a window function.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28291)
<!-- Reviewable:end -->

yuzefovich committed:
sql: return a copy of array in arrayAggregate

Makes arrayAggregate follow its contract by returning
a copy of accumulated array with every call to Result().

Fixes: #28287.

Release note (bug fix): CockroachDB now correctly
handles computation of `array_agg` when used as
a window function.

nvb reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28291)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/builtins/window_builtins.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28291#-LJo2fPJ1-xBxPH-j2m8:-LJo2fPJ1-xBxPH-j2m9:b-f2l9rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd0321489bc2d1578bb28560b3eeed368c16aad/pkg/sql/sem/builtins/window_builtins.go#L297)):*
> ```Go
> }
> 
> // arrayAggForWindow is an alternative implementation of array_agg built-in
> ```

Instead of creating a whole new implementation of array_agg, should we create a wrapper that performs a clone on the result of each call to `Compute`? Also, do any other window function implementations have this problem?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28291)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/builtins/window_builtins.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28291#-LJo2fPJ1-xBxPH-j2m8:-LJoCaru0pwrE6ab6T9k:b-2f7pzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd0321489bc2d1578bb28560b3eeed368c16aad/pkg/sql/sem/builtins/window_builtins.go#L297)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Instead of creating a whole new implementation of array_agg, should we create a wrapper that performs a clone on the result of each call to `Compute`? Also, do any other window function implementations have this problem?
</blockquote></details>

I added a copying wrapper, the code seems cleaner, but I'm not sure how to track the newly allocated memory - we don't have access to `ctx` when creating the wrapper nor when calling `Result` on it.

I think `array_agg` is the only built-in with such problem - all other built-ins return new `Datum` on each new call of `Result`.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
So I think the real problem is that `arrayAggregate` is breaking its contract to never modify datums that it has returned from its Results method. Does the following diff fix the problem?

```
--- a/pkg/sql/sem/builtins/aggregate_builtins.go
+++ b/pkg/sql/sem/builtins/aggregate_builtins.go
@@ -506,7 +506,8 @@ func (a *arrayAggregate) Add(ctx context.Context, datum tree.Datum, _ ...tree.Da
 // Result returns an array of all datums passed to Add.
 func (a *arrayAggregate) Result() (tree.Datum, error) {
        if len(a.arr.Array) > 0 {
-               return a.arr, nil
+               arrCopy := *a.arr
+               return &arrCopy, nil
        }
        return tree.DNull, nil
 }
```

> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28291#-:-LJy95s7EKlK6pLry5Pw:b-2q5hdt)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Yes, that fixes it! Thanks for the investigation!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28291#-:-LJyEz0f3XEJnC6Vgefg:b-nevr2m)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
